Minimum viable implementation of the Kobo sync protocol

shavitmichael opened the pull request:
Introduce API endpoints for the Kobo sync protocol.

Most of the endpoints introduced are trivial placeholders for now, except for three endpoints:
- The v1/library/sync endpoint is used by the device to get a list of new/changed Books from the library since the last sync.
- The v1/library/<book_id>/metadata endpoint is used by the device to obtain a fresh download link for a given book.
- And a third trivial endpoint to get cover images.

This change introduces authentication changes that affects all of WebCalibre. It allows any user to login by passing a valid UserKey token instead of a user/password combination.

To test this change:
- The api_endpoint parameter on the device's .kobo/Kobo/Kobo eReader.conf file must be set to the CalibreWeb server address.
- The new "KoboStore UserKey" value under CalibreWeb's profile configuration should be set to the device's UserKey (found in the user table of .kobo/KoboReader.sqlite). I recommend reading up on the security caution notice in kobo_auth.py before doing so however.
- The new "config server url" value under Server configuration should be set to the server's address (e.g: 192.168.0.1:8083). (I've tried reconstructing this from the request url instead but it's fragile, and doesn't work when the server sits behind a proxy).

The Kobo device will now be able to sync (i.e catalog) all the book in the library, and download them locally for reading. Other features like Shelves, reading stats, etc are not yet implemented.
Known issues:
- This will also prevent the device from syncing to the official Kobo store.

OzzieIsaacs commented on cps/helper.py:
```diff
@@ -428,32 +428,46 @@ def delete_book(book, calibrepath, book_format):
 def get_book_cover(book_id):
     book = db.session.query(db.Books).filter(db.Books.id == book_id).first()
-    if book.has_cover:
+    return get_book_cover_internal(book, False)
```
With use generic cover on failure I get this error:
  File "/usr/local/lib/python3.6/dist-packages/flask/app.py", line 2097, in make_response
    "The view function did not return a valid response. The"
TypeError: The view function did not return a valid response. The function either returned None or ended without a return statement.
Cause are calls to non existing covers, in this case "get_cover_on_failure" returns None. The parameter should be true. Could you please doublecheck this.

shavitmichael replied:
Ah yeah that's correct, in fact that was the whole point of that refactor :).
Fixed.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+from time import gmtime, strftime
+
+from b2sdk.account_info.in_memory import InMemoryAccountInfo
+from b2sdk.api import B2Api
```
To me this looks like remains from the backblaze integration. There is the not called function (get_download_url_for_book_b2) which needs this. From my point of view, this should be be removed including the optional requirement for b2sdk

shavitmichael replied:
Yep, this is a leftover; removed occurrences from this file, optional-requirements and .gitignore.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+    download_authorization = bucket.get_download_authorization(
+        file_path, valid_duration_in_seconds=600
+    )
+    return download_url + "?Authorization=" + download_authorization
```
This function can be removed? Only needed for the backblaze integration

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+
+
+def b64encode_json(json_data):
+    return b64encode(json.dumps(json_data))
```
I had a problem with python3.6. I changed it to (after importing sys on the top):
```
    if sys.version_info < (3, 0):
        return b64encode(json.dumps(json_data))
    else:
        return b64encode(json.dumps(json_data).encode())
```

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+            "NumberFloat": float(book.series_index),
+            # Get a deterministic id based on the series name.
+            "Id": uuid.uuid3(uuid.NAMESPACE_DNS, get_series(book).encode("utf-8")),
+        }
```
This caused an exception on python 3.6.I changed it to:
```
        if sys.version_info < (3, 0):
            name = get_series(book).encode("utf-8")
        else:
            name = get_series(book)
        metadata["Series"] = {
            "Name": get_series(book),
            "Number": book.series_index,
            "NumberFloat": float(book.series_index),
            # Get a deterministic id based on the series name.
            "Id": uuid.uuid3(uuid.NAMESPACE_DNS, name),
        }
```

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+# -*- coding: utf-8 -*-
+
+import base64
+import copy
```
My debugger tells me, this import isn't used, please check

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+import copy
+import uuid
+import os
+from datetime import datetime, tzinfo, timedelta
```
tzinfo and timedelta seem not to be used

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+from flask_login import login_required
+from sqlalchemy import func
+
+from . import config, logger, kobo_auth, ub, db, helper
```
The "ub" import seems not to be used

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+from b2sdk.account_info.in_memory import InMemoryAccountInfo
+from b2sdk.api import B2Api
+from jsonschema import validate, exceptions
+from flask import Blueprint, request, make_response, jsonify, json, send_file
```
send_file seems not to be used

shavitmichael replied:
Done.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,544 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
```
Could you please add here and in kobo_auth.py the GPL3 copyright text like in the other py files

shavitmichael replied:
Done.

OzzieIsaacs commented:
I had some time to take a look at the code. 
I got it running. 
I got the user data from kobo and changed the api endpoint without any problem (this will cause trouble, as my customers are book lovers and not programmers, we need a small script for reading out the user id from the sql file)
I got the device synced (well at least the Kobo is not complaining and seems to be happy).
I added some comments to the code on things I stumbeld over. Could you please check it.

I'm not happy with the server address setting, I'll check it and try to understand it and why it is like it is. 
One thing I didn't like so much: If you enter the userkey, it disappears after reloading the page and you don't see that you entered anything. I know this is also true for the password, but the user key entry now is an optional one. Maybe we should make it readable, or show dummy stars if anything is entered.
Thanks a lot for all the work you put in so far, looks absolutely great.

shavitmichael commented:
Thanks for the review!

> I'm not happy with the server address setting, I'll check it and try to understand it and why it is like it is.

Yeah it's definitely not ideal. I have this commit that works for my set-up without needing the server address setting, I'm just not confident it will work in all scenarios: https://github.com/shavitmichael/calibre-web/commit/ac85b04ee9c988b2d9a4624ded17f91015b01dad .



> I got the user data from kobo and changed the api endpoint without any problem (this will cause trouble, as my customers are book lovers and not programmers, we need a small script for reading out the user id from the sql file)
>
> One thing I didn't like so much: If you enter the userkey, it disappears after reloading the page and you don't see that you entered anything. I know this is also true for the password, but the user key entry now is an optional one. Maybe we should make it readable, or show dummy stars if anything is entered.

To be honest I'm not a big fan of having to enter a userkey in the first place, and I'd rather push for a better solution that doesn't require manually entering the userkey instead of improving this method.
I've documented a few possible alternatives in kobo_auth.py, but here's another idea that doesn't require touching or even looking at the userkey at all:
- Generate a random authentication token for every user and store it in the CalibreWeb user table.
- Add a button in the ui that generates a URL with the token in it, for the user to enter in the api_endpoint device setting (e.g: api_endpoint=....:8083/GENERATED_TOKEN/kobo)
- Although CalibreWeb will still receive the userkey (because we can't prevent the device from sending it), we can ignore it and rely on the CalibreWeb token that is passed in the url instead.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,523 @@
+
+def get_download_url_for_book(book, book_format):
+    return "{url_base}/download/{book_id}/{book_format}".format(
+        url_base=config.config_server_url,
```
I could solve the server name issue by using:

request.environ['werkzeug.request'].base_url 

This worked on with apache as reverse proxy and with direct connection and removes the new config option

shavitmichael replied:
I also discovered Flask.url_for so I've updated all usages I added to that and reverted the config_server_url changes!

shavitmichael replied:
I also discovered Flask.url_for so I've updated all usages I added to that and reverted the config_server_url changes!

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,523 @@
+    # the comparison because of the +00:00 suffix.
+    changed_entries = (
+        db.session.query(db.Books)
+        .filter(func.datetime(db.Books.last_modified) > sync_token.books_last_modified)
```
I found one additional problem. The initial compare will always return no results, as all books in the database are older than the last sync token. So I replaced the > with a !=

shavitmichael replied:
Hmm that sounds like an odd fix: won't this send back all books but the most recently modified ones?
When I was playing around with this a while back, sending "NewEntitlements" for already known uuids would cause the device to invalidate previously downloaded books. Are you not seeing the same issue with "ChangedEntitlements" ?

It is also curious that you ended up in a state where this was needed: the first time the device syncs, its sync_token.books_last_modified is set to datetime.min so that this comparison should pass on the first sync.

OzzieIsaacs replied:
What I did:
Start the first sync, no kepubs in the library -> sync completed, and I guess set the sync date to this date 
Then I debugged the code why no files got downloaded to my reader, or at least the date got set and I found out that the this line of code was hiding all ebooks from the reader. 
I changed the line, started a sync, the epub files list has been send to the kobo reader, but no downloaded (because it's not syncing epubs), and therefore the sync date in the calibre-database wasn't updated.
Afterwards I added one kepub, started the sync it got downloaded and I stopped working.
I think it might be a good idea to add only books with a kepub file to the list of changed books, otherwise the reader always sees a lot of files which it doesn't sync anyway. So we could present an output to the task section of calibre web like: Kobo sync started syncing xx books, ...

shavitmichael replied:
Hmmm I should probably look into this a bit more.
I've added a filter as you suggested in https://github.com/janeczku/calibre-web/pull/1100/commits/d6a974682406edd366cd972b0823002526ec06dc .
I wonder if adding a new format to a book updates that book's last_modified time? (I hope it does...)

shavitmichael replied:
Ok I suspect that between d6a9746, eca4895 and d5dedfb this issue is likely fixed.

OzzieIsaacs commented:
I merged all your changes together with my changes and pushed it to the development branch.

shavitmichael commented:
I haven't had much time to work on this recently, but I'm starting work on replacing the userkey-based authentication with a CalibreWeb generated token as per my last comment.

shavitmichael commented:
I changed the authentication scheme a bit and I think this should make it easier to use. Try visiting http://.../kobo_auth/generate_auth_token :) .

shavitmichael commented:
Ok I'm pretty happy with the current state of this project. At this point the only big features missing are integrations with user BookShelves and Bookmarks.
I've also tested interleaving syncs with book additions as well as a device reset without issue.

Should I squash all these commits together and create a new Pull Request onto Master?

shavitmichael commented:
(FTR: I'll be out on vacation and unlikely to respond for ~2 weeks).

I've run into a subtle but annoying limitation with [FixedLayout epubs](https://www.w3.org/publishing/epub3/epub-packages.html#fxl-intro).
When a book is imported over USB, the opf is parsed to figure out whether it should be rendered using a FixedLayout (by checking for _rendition:layout=pre-paginated_ as per the spec).
When a book is imported from a DownloadUrl, the rendition:layout attribute is ignored in favor of the DownloadUrl's  _format_ attribute.

Permutations I've tried so far:
-   Format KEPUB or EPUB3
-> _rendition:layout=pre-paginated_ isn't respected.
-  Format FLEPUB or EPUB3FL
-> In this case we have the opposite issue and _rendition:layout=reflowable_ isn't respected.

I've tried reverse engineering libnickel, and it looks like there's a code-path where a downloaded book can enable FixedLayouts when the epubType is set to -1. From trial and error however, it doesn't look the device will download formats it doesn't recognize (and those that it recognizes end-up setting an epubType > 0).

We may have to parse epubs are the CalibreWeb side to determine if they're FixedLayout and whether we should send the EPUB3 or EPUB3FL format back to the user; but that's probably something to fix another time.

OzzieIsaacs commented:
I had also a little internet holiday, so here is my feedback:
I encounter one problem. You hijacked the RemoteAuthToken table, this table is always deleted if you hit the submit button in the basic config section with deactivated "magic link". This leads to the problem that upon doing so, the stored kobo token is deleted. 
- [ ] I suggest to add a coloum 'token_type' or so (to be prepared for other types of token to store in the table. Then it would be possible to delete the "normal" magic link tokens independent from the kobo tokens.

- [ ] I also think we need a button in the config section to generate the kobo tokens and also to invalidate them (in case you device gets stolen, or lost). The token should also be visible in the user section, so you can read the key later on, if you accidentally closed the key displaying window (optional?).

- [ ] I also think the kobo syncing should be optional (currently it adds the jsonschema dependency, and without it, it fails, which is not backwards compatible). 

- [ ] I'd like to place it together with the ldap and goodreads integration in the services folder

- [ ] I also like to have version info from all dependencies in the about section, for better debugging informations from users.

Don't panic, I'm absolutly willing to help implementing my recommendations. Currently (hopefully only till sunday) I busy with another project, so I can start implementing on this topics.

OzzieIsaacs commented:
I did the proposed changes in the development branch. I have to do some tests, than we can merge it into master

OzzieIsaacs commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 15
           

Complexity increasing per file
==============================
- cps/kobo.py  6
- cps/kobo_auth.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308&bid=16085444)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+
+            data_json = sync_token_json["data"]
+            validate(sync_token_json, SyncToken.data_schema_v1)
+        except (exceptions.ValidationError, ValueError) as e:
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'e'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'e' is assigned to but never used](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+
+def get_store_url_for_current_request():
+    # Programmatically modify the current url to point to the official Kobo store
+    base, sep, request_path_with_auth_token = request.full_path.rpartition("/kobo/")
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'sep'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

shavitmichael replied:
I personally prefer leaving these named since they document the url format and rpartition return values.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+    @staticmethod
+    def from_headers(headers):
+        sync_token_header = headers.get(SyncToken.SYNC_TOKEN_HEADER, "")
+        if sync_token_header == "":
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Possible hardcoded password: ''](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo_auth.py:
```diff
@@ -0,0 +1,133 @@
+
+def register_url_value_preprocessor(kobo):
+    @kobo.url_value_preprocessor
+    def pop_auth_token(endpoint, values):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'pop_auth_token'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+        books_last_modified: Datetime representing the last modified book that the device knows about.
+    """
+
+    SYNC_TOKEN_HEADER = "x-kobo-synctoken"
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Possible hardcoded password: 'x-kobo-synctoken'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

shavitmichael replied:
See other comment response about the hardcoded password check.

OzzieIsaacs commented on cps/kobo_auth.py:
```diff
@@ -0,0 +1,133 @@
+
+Log-in:
+When first booting a Kobo device the user must sign into a Kobo (or affiliate) account.
+Upon successful sign-in, the user is redirected to 
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Trailing whitespace](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo_auth.py:
```diff
@@ -0,0 +1,133 @@
+    <script type='text/javascript'>location.href='kobo://UserAuthenticated?userId=<redacted>&userKey<redacted>&email=<redacted>&returnUrl=https%3a%2f%2fwww.kobo.com';</script>.
+And triggers the insertion of a userKey into the device's User table.
+
+Together, the device's DeviceId and UserKey act as an *irrevocable* authentication 
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Trailing whitespace](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+#  You should have received a copy of the GNU General Public License
+#  along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import os
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['os' imported but unused](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused import os](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+        },
+    }
+
+    def __init__(
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Possible hardcoded password: ''](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

shavitmichael replied:
I'm not sure we should worry too much about this sort of false-positive 'harcoded password' check.

OzzieIsaacs replied:
It's no problem, maybe I should deactivate the check in codacy.

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+    from urllib.parse import unquote
+
+from jsonschema import validate, exceptions
+from flask import (
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['flask.json' imported but unused](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused json imported from flask](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,508 @@
+def get_store_url_for_current_request():
+    # Programmatically modify the current url to point to the official Kobo store
+    base, sep, request_path_with_auth_token = request.full_path.rpartition("/kobo/")
+    auth_token, sep, request_path = request_path_with_auth_token.rstrip("?").partition(
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'auth_token'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

shavitmichael replied:
I personally prefer leaving these named since they document the url format.

OzzieIsaacs commented on cps/kobo_auth.py:
```diff
@@ -0,0 +1,133 @@
+Upon successful sign-in, the user is redirected to 
+    https://auth.kobobooks.com/CrossDomainSignIn?id=<some id>
+which serves the following response:
+    <script type='text/javascript'>location.href='kobo://UserAuthenticated?userId=<redacted>&userKey<redacted>&email=<redacted>&returnUrl=https%3a%2f%2fwww.kobo.com';</script>.
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [line too long (176 > 159 characters)](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented:
I found 2 new possible problems:
In case you change your calibre-web library and try to delete the books on the reader afterwards, all delete requests are proxied to kobo, which answers with error. This causes the reader upon the next sync to try to delete the books again with the same result. I think we should answer this request with "ok".
If you try to create the apiendpoint entry, by accessing you server via localhost or 127.0.0.1, this will be written into the apiendpoint string. Of course this will not work later on. I this we should present a warning text unter the apipoint text, as this can't be avoided.

shavitmichael commented:
Wow thanks a lot for the help on this integration! I haven't yet had a chance to fully catch up on your changes but I'll likely take a look later this week.

I might also start implementing some of the other API calls like shelves/collections or book deletions (to address your latest comment).

OzzieIsaacs commented:
I had a big problem with my kobo reader: It was requesting urls in form http:/[url_of_server]/kobo..., but the server is running on port 8083. The request was send to the right port, but with was missing the port number. http:/[url_of_server]:[Port_number]/kobo.... This caused the function url_for to generate false urls (without the port number), so the kobo device was requesting books and covers from a non existing server on port :80. I could fix this now. I can now download epub books and also see the covers (which never worked before, but I didn't get why). And your fix regarding the sync time worked, I now can get all books with the sync request.

OzzieIsaacs commented:
Ah: And I added a switch preventing kobo from phoning home, which I don't like, because I never use the Kobo store.
The also added enable kobo function is currently not working, it's always activated if all dependencies are fulfilled.

OzzieIsaacs commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 66
           

Complexity increasing per file
==============================
- cps/services/SyncToken.py  6
- cps/kobo.py  6
- cps/ub.py  1
- cps/static/js/table.js  1
- cps/kobo_auth.py  3
- cps/db.py  2
- cps.py  1
         

Clones added
============
- cps/admin.py  41
- cps/shelf.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308&bid=16085444)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+                type: 'Post',
+                data: $(this).closest("form").serialize() + "&" + $(this)[0].name + "=",
+                success: function () {
+                $.ajax ({
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expected indentation of 20 spaces but found 16.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+            return;
+        });
+    }
+    $('#get_column_values').on('click',function()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [A space is required after ','.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+            // event.stopPropagation();
+            // event.preventDefault();
+            $(this)[0].blur();
+            console.log($(this)[0].name);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unexpected console statement.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+            },
+            url: path + "/../../ajax/listrestriction/" + type,
+            rowStyle: function(row, index) {
+                console.log('Reihe :' + row + ' Index :'+ index);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+            },
+            onClickCell: function (field, value, row, $element) {
+                if(field == 3){
+                    console.log("element")
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unexpected console statement.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+    $('#restrictModal').on('hidden.bs.modal', function () {
+        // Destroy table and remove hooks for buttons
+        $("#restrict-elements-table").unbind();
+        $('#restrict-elements-table').bootstrapTable('destroy');
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo.py:
```diff
@@ -0,0 +1,386 @@
+
+import sys
+import uuid
+from datetime import datetime
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused datetime imported from datetime](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+                    method:"get",
+                    url: path + "/../../ajax/listrestriction/"+type,
+                    async: true,
+                    timeout: 900,
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expected indentation of 24 spaces but found 20.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/admin.py:
```diff
@@ -247,6 +252,229 @@ def list_domain(allow):
+@admi.route("/ajax/addrestriction/<int:type>", methods=['POST'])
+@login_required
+@admin_required
+def add_restriction(type):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'type'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+                    url: path + "/../../ajax/listrestriction/"+type,
+                    async: true,
+                    timeout: 900,
+                    success:function(data) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expected indentation of 24 spaces but found 20.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/main.js:
```diff
@@ -228,6 +228,41 @@ $(function() {
+
+    $("#btndeletetoken").click(function() {
+        //get data-id attribute of the clicked element
+        var pathname = document.getElementsByTagName("script"), src = pathname[pathname.length-1].src;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Operator '-' must be spaced.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+                }
+            },
+            onClickCell: function (field, value, row, $element) {
+                if(field == 3){
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Missing space before opening brace.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expected '===' and instead saw '=='.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+                    return {classes: 'bg-dark-danger'}
+                }
+            },
+            onClickCell: function (field, value, row, $element) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['$element' is defined but never used.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+                    timeout: 900,
+                    success:function(data) {
+                        $("#restrict-elements-table").bootstrapTable("load", data);
+                       }
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expected indentation of 24 spaces but found 23.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/main.js:
```diff
@@ -228,6 +228,41 @@ $(function() {
+        })
+        .on("hidden.bs.modal", function() {
+            $(this).find(".modal-body").html("...");
+             $("#config_delete_kobo_token").show();
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Expected indentation of 12 spaces but found 13.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+        $('#h4').addClass('hidden');
+    });
+    function startTable(type){
+        var pathname = document.getElementsByTagName("script"), src = pathname[pathname.length-1].src;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Operator '-' must be spaced.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+            striped: false
+        });
+        $("#restrict-elements-table").removeClass('table-hover');
+        $("#restrict-elements-table").on('editable-save.bs.table', function (e, field, row, old, $el) {
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+            console.log($(this)[0].name);
+            $.ajax({
+                url: path + "/../../ajax/addrestriction/"+type,
+                type: 'Post',
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/admin.py:
```diff
@@ -247,6 +252,229 @@ def list_domain(allow):
+            elementlist[int(element['id'][1:])]=element['Element']
+            config.config_restricted_column_value = ','.join(elementlist)
+            config.save()
+            pass
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unnecessary pass statement](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/services/SyncToken.py:
```diff
@@ -0,0 +1,148 @@
+        books_last_modified: Datetime representing the last modified book that the device knows about.
+    """
+
+    SYNC_TOKEN_HEADER = "x-kobo-synctoken"
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Possible hardcoded password: 'x-kobo-synctoken'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/admin.py:
```diff
@@ -247,6 +252,229 @@ def list_domain(allow):
+@admi.route("/ajax/listrestriction/<int:type>")
+@login_required
+@admin_required
+def list_restriction(type):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'type'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/static/js/table.js:
```diff
@@ -93,6 +93,116 @@ $(function() {
+        $('#h2').removeClass('hidden');
+    });
+
+    $('#get_tags').on('click',function()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/kobo_auth.py:
```diff
@@ -0,0 +1,147 @@
+
+def register_url_value_preprocessor(kobo):
+    @kobo.url_value_preprocessor
+    def pop_auth_token(endpoint, values):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Unused variable 'pop_auth_token'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented on cps/admin.py:
```diff
@@ -247,6 +252,229 @@ def list_domain(allow):
+@admi.route("/ajax/editrestriction/<int:type>", methods=['POST'])
+@login_required
+@admin_required
+def edit_restriction(type):
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Redefining built-in 'type'](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846308)

OzzieIsaacs commented:
I had some time to do some more tests. Your  login manger code was conflicting with the opds login, so I modified it a bit (it's now no longer login decorators, but needs_kobo_authentication). At this point I realized that preventing the proxiing to kobo prevents the inital /auth/device requests and the generation of the bearer token. I changed this, so this request is proxied again. But my goal is still to remove all home calls.

shavitmichael commented:
> I had some time to do some more tests. Your login manger code was conflicting with the opds login, so I modified it a bit (it's now no longer login decorators, but needs_kobo_authentication). At this point I realized that preventing the proxiing to kobo prevents the inital /auth/device requests and the generation of the bearer token. I changed this, so this request is proxied again. But my goal is still to remove all home calls.

Before I added proxiying to Kobo's servers, this PR previously had a dummy implementation for /auth/device [here ](https://github.com/janeczku/calibre-web/blob/53578671033d57c62835fb863b80d8414a66a771/cps/kobo.py#L431) :) .

Speaking of, Kobo servers were down[ earlier today](https://goodereader.com/blog/electronic-readers/kobo-website-and-apps-are-down) which prevented me from syncing CalibreWeb books. I'm thinking of adding timeouts on proxied requests, which could fallback to the dummy implementations when necessary (e.g: /auth/device).

OzzieIsaacs commented:
I merged the PR to the master branch.
The open topic regarding the formating kepub/epub should be solved in a new PR (if possible). One big task is pending: creating documentation

shavitmichael commented:
Thanks Ozzie, that's awesome!!

------

FTR, I've created a few more issues to better document ongoing or needed efforts:
- Open Issue #1219 - to track the FixedLayout epub problem.
- PR #1164 - implementing book archiving.
- Draft PR #1220 - implementing ReadingState syncing.
